### ISSUE TITLE
Add `SHELL_PROMPT_PREFIX` et al to default prompts

### ISSUE THREAD

**Issue by septatrix** [fish-shell/fish-shell#10924]
The most recent systemd version has introduced a set of new standard environment variables  that shall be consumed by shells for their prompt. These are [`SHELL_PROMPT_PREFIX`, `SHELL_PROMPT_SUFFIX`, and `SHELL_WELCOME`](https://www.freedesktop.org/software/systemd/man/257/pam_systemd.html#%24SHELL_PROMPT_PREFIX).

> They are passed to the invoked session processes, where they are imported into any shell prompt (specifically `$SHELL_PROMPT_PREFIX` is added as prefix to `$PS1`, and `$SHELL_PROMPT_SUFFIX` as suffix) or printed on screen when a shell first initializes.

They are for example also [set when using `run0`](https://www.freedesktop.org/software/systemd/man/257/run0.html#--shell-prompt-prefix=STRING) (an suid-less alternative to sudo/doas) and there have been [ideas to also set them when entering containers, VMs, remote hosts etc](https://github.com/systemd/systemd/pull/34297).

Systemd itself provides a profile.d file for this purpose but these are not used by fish. It would therefore be great if fish picked them up in its default prompts.

More details in the initial PR: https://github.com/systemd/systemd/pull/34297

**Comment by mqudsi** [comment]
We've done things to remove pollution of the shell from the environment in the past. However I could see how this change in particular could be seen as more help than hinderance.

Wouldn't the right place for this be in distro-maintained patches against upstream fish? i.e. systemd-using distros would patch the default prompts for fish installed via their repos to include systemd cruft rather than polluting the upstream project, which has no systemd dependencies, runs on platforms systemd doesn't support, etc, etc

**Comment by krobelus** [comment]
If they advertise it as non-systemd specific standard then we should implement it, but I'm not sure if that's the case.
Since systemd already supports other shells, it wouldn't be out of line for it to support fish too (adding to /usr/share/fish/vendor_conf.d/).

From a fish perspective it might seem odd that the variables are static strings and don't allow arbitrary computation whenever a prompt is drawn.
I guess it is powerful enough for current use cases.

**Comment by septatrix** [comment]
> Wouldn't the right place for this be in distro-maintained patches against upstream fish?

If possible it is preferred to avoid downstream patching. Some distros (esp. Fedora) have a very strong upstream-first mantra.

Another annoying thing is that this would need patching in several files instead of only a single function like `fish_prompt_prefix`. If there is a simple way to do this by adding a file in e.g. `/usr/share/fish/vendor_conf.d` which I am missing that would be great.

> If this goes from being a systmed thing to a universal posixy environment de facto standard, that would be a different story.

If systemd implements this for machinectl/nspawn these would also be set in containers where systemd is not present. This would also be a great and simple way to generally prepend/append something to the prompt. E.g. `ssh -o SetEnv=SHELL_PROMPT_PREFIX="foobar" server`.

**Comment by septatrix** [comment]
> If they advertise it as non-systemd specific standard then we should implement it, but I'm not sure if that's the case.

They intend it to be generally used by other projects as well but given that it just got released there are no other users of this yet. Similar to their container interface which e.g. tells container managers to set `$container=podman` etc or `$LISTEN_FD` spec which are now also used by other projects.

> Since systemd already supports other shells, it wouldn't be out of line for it to support fish too (adding to /usr/share/fish/vendor_conf.d/).

Yeah that would work. If we want to keep it to only systemd systems for now I guess systemd folks might also be fine with adding it to their repo - they already did for their `profile.d` file which adds these envs to `PS1`. But can simply config files easily append/prepend stuff to prompts defined in themes? I am not sure in which order files are loaded and at which point to prompt gets set. I am under the assumption that the vendor configs get loaded first and the prompt is set later so we cannot wrap that function?

> From a fish perspective it might seem odd that the variables are static strings and don't allow arbitrary computation whenever a prompt is drawn. I guess it is powerful enough for current use cases.

Yeah for the limited scope they are perfect. Stuff like displaying a container name or a Unicode symbol which are all known by the process setting those environment variable.

**Comment by krobelus** [comment]
>  But can simply config files easily append/prepend stuff to prompts defined in themes?

yeah that might be hacky because `fish_prompt` is the single source of truth.
We should make something that doesn't rely on load order.
We could either support `SHELL_PROMPT_PREFIX` directly (hardcoded)
or add a new function and systemd would do `function fish_prompt_prefix; echo $SHELL_PROMPT_PREFIX; end`.
The indirection doesn't make as much sense if it's used by anything other than systemd.
I think one question is, how will integration (for other shells) look like; would everyone copy systemd's profile.d entry and try not to add it if it's already there?

**Comment by faho** [comment]
We already include the virtual env name in a bunch of our prompts,
so I don't think adding something that is currently only used by systemd but agnostic in principle is an issue.

**Comment by zanchey** [comment]
As long as no-one does anything clever with expecting special characters or command substitutions to work...

**Comment by mqudsi** [comment]
> As long as no-one does anything clever with expecting special characters or command substitutions to work...

This is a very good point. I wonder if it's worth running it through some sanitization function to strip any non-static strings from the payload.

Searching GitHub, I can't actually find much in-the-wild usage of it (unless they're not indexed for some reason) besides a few random plaintext usages: https://github.com/search?q=%22SHELL_PROMPT_PREFIX%22&type=code&p=1

**Comment by thomasbbrunner** [comment]
> or add a new function and systemd would do function fish_prompt_prefix; echo $SHELL_PROMPT_PREFIX; end.

Something like this would also make sense for people that want to customize the default prompt. Currently, this would require copying the default `fish_prompt` and modifying, but IMO this could be easier.

For instance, the function could either make use of the env vars:

```
function fish_prompt
  set -l prompt ...
  echo $SHELL_PROMPT_PREFIX$prompt$SHELL_PROMPT_SUFFIX
end
```

Or the function could take (optional) arguments:

```
fish_prompt suffix prefix
```

This way, users could easily customize the default prompt by specifying env vars or by wrapping the `fish_prompt` function and passing arguments:

```
functions -c fish_prompt _old_fish_prompt

function fish_prompt
  set -l prefix ...  
  set -l suffix ...  
  echo _old_fish_prompt $prefix $suffix
end
```

Happy to assist in implementing this if there's interest.

**Comment by krobelus** [comment]
> Happy to assist in implementing this if there's interest.

I think supporting SHELL_PROMPT_PREFIX out of the box would be a good start.
We can add more flexible knobs later, if really needed.

**Comment by edouard-lopez** [comment]
For those wondering what it looks like, here is a screenshot of the rendering using [pure-fish/docker-fish](https://github.com/pure-fish/docker-fish) image:

<img width="660" height="234" alt="usage of SHELL_PROMPT_PREFIX, SHELL_PROMPT_SUFFIX, and SHELL_WELCOME variables in Fish shell" src="https://github.com/user-attachments/assets/ba1cd9a1-66b5-4c68-bbcf-61381f469dd2" />
